### ISSUE TITLE
Can the rule `no-use-before-define` be smater in such a case?

### ISSUE THREAD

**Issue by aleen42** [eslint/eslint#14080]
### Environments

* **ESLint Version:** v7.16.1
* **Node Version:** 6.14.4
* **npm Version:** 13.14.0

### Descipritions

It seems like the rule (`no-use-before-define`) is not so smart that it can warn that `b` will probably be used before it is defined when the condition `a` matches.

```js
function test() {
    const a = 1;
    if (a) {
        console.log(b); // ESLint: 'b' was used before it was defined.(no-use-before-define)
        setTimeout(function () {
            console.log(b); // OK
        });
        return;
    }
    const b = 1;
}
```

**Comment by BairDev** [comment]
Is there any reason for not putting `const b = 1;` right after `const a = 1;`? Do you want to save some work for the garbage collector? I think that initializing the constants would be much more readable.

**Comment by aleen42** [comment]
I know it is more readable, but this was the case when I hoped to keep the historical code by not moving it.

**Comment by nzakas** [comment]
The rule is working as intended. If you’d like to keep the format of your current code, you should disable the rule either globally or just in that code by using a disable comment.
